gapis/trace/android/adreno/validate.go

```diff
@@ -0,0 +1,131 @@
+// Copyright (C) 2019 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package adreno
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/google/gapid/core/log"
+	"github.com/google/gapid/gapis/perfetto"
+	perfetto_service "github.com/google/gapid/gapis/perfetto/service"
+	"github.com/google/gapid/gapis/trace/tracer"
+)
+
+const (
+	counterIdQuery = "" +
+		"select counter_id from counter_definitions " +
+		"where name = '%v'"
+	counterValuesQuery = "" +
+		"select value from counter_values " +
+		"where counter_id = %v order by ts " +
+		"limit %v offset 10"
+	sampleCounter = 100
+)
+
+var (
+	counters = []tracer.GpuCounter{
+		{1, "Clocks / Second", checkLargerThanZero},
+		{4, "GPU % Bus Busy", checkLargerThanZero},
+		{20, "% Shaders Busy", checkLargerThanZero},
+	}
+)
+
+type AdrenoValidator struct {
+}
+
+type checker func(column *perfetto_service.QueryResult_ColumnValues, columnType perfetto_service.QueryResult_ColumnDesc_Type) bool
+
+func checkLargerThanZero(column *perfetto_service.QueryResult_ColumnValues, columnType perfetto_service.QueryResult_ColumnDesc_Type) bool {
+	longValues := column.GetLongValues()
+	doubleValues := column.GetDoubleValues()
+	for i := 0; i < sampleCounter; i++ {
+		if columnType == perfetto_service.QueryResult_ColumnDesc_LONG {
+			if longValues[i] <= 0 {
+				return false
+			}
+		} else if columnType == perfetto_service.QueryResult_ColumnDesc_DOUBLE {
+			if doubleValues[i] <= 0.0 {
+				return false
+			}
+		}
+	}
+	return true
+}
+
+func check(ctx context.Context, result *perfetto_service.QueryResult, f checker) bool {
+	columnDescriptor := result.GetColumnDescriptors()[0]
+	column := result.GetColumns()[0]
+	columnType := columnDescriptor.GetType()
+	if columnType != perfetto_service.QueryResult_ColumnDesc_LONG && columnType != perfetto_service.QueryResult_ColumnDesc_DOUBLE {
+		return false
+	}
```

Comment:
make this a `isNumber` checker, then create a `func add(c1, c2 checker) checker` and use `and(isNumber, isGreaterThanZero)`, etc.